cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i

```diff
@@ -0,0 +1,304 @@
+            {
+                classWeights[j] = 0;
+            }
+            for (size_t j = 0; j < k; ++j)
+            {
+                classWeights[trainLabel[indices[i * k + j]]] += 1;
+            }
+            size_t maxWeightClass = 0;
+            size_t maxWeight      = 0;
+            for (size_t j = 0; j < nClasses; ++j)
+            {
+                if (classWeights[j] > maxWeight)
+                {
+                    maxWeight      = classWeights[j];
+                    maxWeightClass = j;
+                }
+            }
+            testLabel[i] = maxWeightClass;
+        }
+        return services::Status();
+    }
+
+    services::Status distanceWeightedVoting(size_t nClasses, size_t k, size_t n, FPType * distances, int * indices, const int * trainLabel,
+                                            int * testLabel)
+    {
+        daal::services::internal::TArray<FPType, cpu> classWeightsArr(nClasses);
+        FPType * classWeights = classWeightsArr.get();
+        DAAL_CHECK_MALLOC(classWeights);
+
+        const FPType epsilon = daal::services::internal::EpsilonVal<FPType>::get();
+        bool isContainZero   = false;
+        for (size_t i = 0; i < k * n; ++i)
+        {
+            if (distances[i] < epsilon && distances[i] > -epsilon)
+            {
+                isContainZero = true;
+                break;
+            }
+        }
+
+        for (size_t i = 0; i < n; ++i)
+        {
+            for (size_t j = 0; j < nClasses; ++j)
+            {
+                classWeights[j] = 0;
+            }
+            for (size_t j = 0; j < k; ++j)
+            {
+                if (isContainZero)
+                {
+                    if (distances[i] < epsilon && distances[i] > -epsilon)
+                    {
+                        classWeights[trainLabel[indices[i * k + j]]] += 1;
+                    }
+                }
+                else
+                {
+                    classWeights[trainLabel[indices[i * k + j]]] += 1 / distances[i * k + j];
+                }
+            }
+            size_t maxWeightClass = 0;
+            FPType maxWeight      = 0;
+            for (size_t j = 0; j < nClasses; ++j)
+            {
+                if (classWeights[j] > maxWeight)
+                {
+                    maxWeight      = classWeights[j];
+                    maxWeightClass = j;
+                }
+            }
+            testLabel[i] = maxWeightClass;
+        }
+        return services::Status();
+    }
+
+    services::Status computeDistances(const NumericTable * trainTable, const NumericTable * testTable)
+    {
+        daal::SafeStatus s;
+
+        const size_t nDims  = trainTable->getNumberOfColumns();
+        const size_t nTrain = trainTable->getNumberOfRows();
+        const size_t nTest  = testTable->getNumberOfRows();
+
+        _distances = static_cast<FPType *>(services::internal::service_malloc<FPType, cpu>(nTrain * nTest * sizeof(FPType)));
```

Comment:
you don't need to keep all this quite large block of data: ntrain x ntest. It's non-efficient in terms of memory capacity and performance.

you can process only 128 x 128 block in one parallel task:
```py
parallel_for out_block in out_blocks:
   seq for in_block in in_blocks:
         dist_128x128 = compute_distance(out_block, in_block);
         update_knn(dist_128x12);
   sort_knn_by_distances(out_block);
   if (is_classification):
         do_classifcation();
```
So, you will have only one parallel_for in the algorithm, you don't need to store all distances and will have optimal perf

Reply:
I changed algorithm flow a bit according to comment, full appliance of it will be in future optimizations